pkg/service/new_reconciler.go

```diff
@@ -0,0 +1,152 @@
+			}
+			params := reconciler.Params[*loadbalancer.Service]{
+				Config:    p.Config,
+				Lifecycle: p.Lifecycle,
+				Log:       p.Log,
+				DB:        p.DB,
+				Jobs:      p.Jobs,
+				Metrics:   p.Metrics,
+				ModuleId:  p.ModuleId,
+				Scope:     p.Scope,
+				Table:     s.UnsafeRWTable(),
+			}
+			return reconciler.Register(params)
+		},
+	),
+)
+
+// Temporary workaround as the RWTable[*Service] cannot be provided even privately
+// when another module has it already provided privately. Will need to change the
+// reconciler to take the Table as part of Config instead.
+// https://github.com/cilium/statedb/pull/20
+type reconcilerParamsPartial[Obj comparable] struct {
+	cell.In
+
+	Config    reconciler.Config[Obj]
+	Lifecycle cell.Lifecycle
+	Log       logrus.FieldLogger
+	DB        *statedb.DB
+	Jobs      job.Registry
+	Metrics   *reconciler.Metrics
+	ModuleId  cell.ModuleID
+	Scope     cell.Scope
+}
+
+func newServicesReconcilerConfig(p serviceManagerOpsParams) reconciler.Config[*loadbalancer.Service] {
+	ops := &serviceManagerOps{p}
+	return reconciler.Config[*loadbalancer.Service]{
+		FullReconcilationInterval: 0, // No full reconciliation for now.
+		RetryBackoffMinDuration:   100 * time.Millisecond,
+		RetryBackoffMaxDuration:   time.Minute,
+		IncrementalRoundSize:      300,
+		GetObjectStatus:           (*loadbalancer.Service).GetStatus,
+		WithObjectStatus:          (*loadbalancer.Service).WithStatus,
+		RateLimiter:               rate.NewLimiter(5*time.Millisecond, 1),
+		Operations:                ops,
+		BatchOperations:           nil,
+	}
+}
+
+type serviceManagerOpsParams struct {
+	cell.In
+
+	Backends statedb.Table[*loadbalancer.ServiceBackend]
+	Manager  ServiceManager
+}
+
+type serviceManagerOps struct {
+	serviceManagerOpsParams
+}
+
+// Delete implements reconciler.Operations.
+func (ops *serviceManagerOps) Delete(ctx context.Context, txn statedb.ReadTxn, svc *loadbalancer.Service) error {
+	_, err := ops.Manager.DeleteService(svc.Frontend)
+	return err
+}
+
+// Prune implements reconciler.Operations.
+func (s *serviceManagerOps) Prune(context.Context, statedb.ReadTxn, statedb.Iterator[*loadbalancer.Service]) error {
+	// We cannot implement Prune() since this is an additional interface towards ServiceManager and thus not
+	// the only source. ServiceManager itself does the pruning when SyncWithK8sFinished() is invoked.
+	return nil
+}
+
+// Update implements reconciler.Operations.
+func (ops *serviceManagerOps) Update(ctx context.Context, txn statedb.ReadTxn, svc *loadbalancer.Service, changed *bool) error {
+	// Collect all backends for the service with matching protocols (L3&L4) and convert
+	// to the SVC type.
+	iter := loadbalancer.GetBackendsForService(txn, ops.Backends, svc)
+	bes := statedb.Collect(
+		statedb.Filter(iter, func(be *loadbalancer.ServiceBackend) bool {
+			return svc.Frontend.ProtoEqual(&be.L3n4Addr)
+		}))
+	_, _, err := ops.Manager.UpsertService(toSVC(svc.Frontend, svc, bes))
```

Comment:
How are stale endpoints pruned? Not seeing any evidence of that happening here, but I'm probably just missing something obvious.

Reply:
`ServiceManager` keeps track of backends that are referenced by services and removes the ones that do not exist. The UpsertService call here tells ServiceManager about all the backends currently associated to a service.

I would like to replace `ServiceManager` later on with a StateDB reconciler that takes the `Service` and queries for associated backends and similarly figures out which backends are now gone and need to be removed from the backends BPF map.

Looks roughly like this: https://github.com/cilium/cilium/pull/31755/files#diff-1aab408d0c33a6dfa08997f13c3a94f372eb95a331b731c09752240a960be346R45